File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/service/transactions/base/UaTransaction2.java
package org.opcfoundation.webserver.service.transactions.base;

import org.opcfoundation.webapi.service.types.ServiceContext;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;

public abstract class UaTransaction2<T_Request, T_Response> {
    protected final ServiceContext serviceContext;
    private final List<Integer> handleIds;

    public UaTransaction2(
            ServiceContext serviceContext,
            List<Integer> handleIds)
    {
        this.serviceContext = serviceContext;
        this.handleIds = handleIds;
    }

    public ServiceContext getServiceContext() {
        return serviceContext;
    }

    public List<Integer> getHandleIds() {
        return handleIds;
    }

    public List<T_Request> getRequestedItems()
    {
        List<T_Request> items = new ArrayList<>();
        List<T_Request> allItems = getItems();

        for (Integer item : handleIds)
        {
            items.add(allItems.get(item));
        }

        return items;
    }

    public abstract List<T_Request> getItems();

    public abstract List<T_Response> getResults();

    public abstract CompletableFuture<Void> execute();
}

File: opcua-webservice-node/source/src/server/service/transactions/base/UaTransaction2.ts
import { ServiceContext } from "../../../types/contexts";

export abstract class UaTransaction2<TRequest, TResponse> {
    protected readonly _serviceContext: ServiceContext;
    private readonly _handleIds: Array<number>;

    constructor(serviceContext: ServiceContext, handleIds: Array<number>) {
        this._serviceContext = serviceContext;
        this._handleIds = handleIds;
    }

    get serviceContext(): ServiceContext {
        return this._serviceContext;
    }

    get handleIds(): Array<number> {
        return this._handleIds;
    }

    getRequestedItems(): Array<TRequest> {
        const allItems = this.getItems();
        let requestedItems: Array<TRequest> = [];

        for (const handleId of this._handleIds) {
            if (handleId < 0 || handleId >= allItems.length) continue;
            requestedItems.push(allItems[handleId]);
        }

        return requestedItems;
    }

    abstract getItems(): Array<TRequest>;

    abstract getResults(): TResponse[];

    abstract execute(): Promise<void>;
}